File: Array/MaxIndex.cpp
#include<bits/stdc++.h>
using namespace std;

// logic: two pointer approach 

int maxIndexDiff(vector<int>& arr) {
        int n = arr.size();
         // Step 1: Create and fill leftMin array
        vector<int> lMin(n);
        lMin[0] = arr[0];
        for (int i = 1; i < n; i++) {
            lMin[i] = min(arr[i], lMin[i - 1]);
        }
    
        // Step 2: Create and fill rightMax array
        vector<int> rMax(n);
        rMax[n - 1] = arr[n - 1];
        for (int j = n - 2; j >= 0; j--) {
            rMax[j] = max(arr[j], rMax[j + 1]);
        }
    
        // Step 3: Traverse both arrays using two pointers
        int i = 0, j = 0, maxDiff = -1;
        while (i < n && j < n) {
            if (lMin[i] <= rMax[j]) {
                maxDiff = max(maxDiff, j - i);
                j++;
            } else {
                i++;
            }
        }
    
        return maxDiff;
    }


int main() {
    vector<int> arr = {15};
    int n = arr.size();
    cout << maxIndexDiff(arr) << endl;
    return 0;
}

File: Array/MaxIndex.py
def maxIndexDiff(arr):
    n = len(arr)

    # left min
    lMin = [0] * n
    lMin[0] = arr[0]
    for i in range (1,n):
        lMin[i] = min(arr[i], lMin[i-1])

    # right max
    rMax = [0] * n 
    rMax[n-1] = arr[n-1]
    for j in range(n - 2, -1, -1):
        rMax[j] = max(arr[j], rMax[j + 1])

    i, j, maxDiff = 0, 0, -1
    while i < n and j < n:
        if lMin[i] <= rMax[j]:
            maxDiff = max(maxDiff, j-i)
            j +=1
        else:
            i +=1
            
    return maxDiff


arr = [15]
result = maxIndexDiff(arr)
print(result)